### PR TITLE
Mask dev

### PR DESCRIPTION
**Pull request by magska**

### CODE REVIEW

File: buildscript/buildFiles.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	mkdirp = require('mkdirp'),
+	Validate = require('./validate');
+
+function Build() {
+	this.configPath = './configs/';
+	this.fs = fs;
+	this.mkdirp = mkdirp;
+}
+
+Build.prototype.doBuild = function() {
```
Comment (rob-mccann) [inline buildscript/buildFiles.js]: can we use grunt for any of these things rather than writing ourselves?

File: configs/blocket.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+    "client": "blocket",
+    "config": [
+        {
+            "throttle": 1,
+			"dataCollector": "https://collector.schibsted.io/api/v1/track",
```
Comment (rob-mccann) [inline configs/blocket.json]: indentation

File: buildscript/buildFiles.js (hunk 1 above)
Comment (magska) [inline buildscript/buildFiles.js]: We probably could, but this script will be more advanced in the future (with plugins and stuff), so decided to write it myself. Also, that was the most fun alternative.

File: test/buildscript-test.js [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+ So we use the karma plugin sinon-chai to inject them as globals as a workaround
+ */
+/*globals sinon, expect*/
+'use strict';
+
+var Build = require('../buildscript/buildFiles');
+var build = new Build();
+
+var okConfigs = {
+	blocket: {
```
Comment (rob-mccann) [inline test/buildscript-test.js]: do we need all of these? Can they go before the tests they pertain to?

File: test/buildscript-test.js [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+			}
+		]
+	}
+};
+
+describe('buildscript', function() {
+	describe('getAllConfigFiles', function() {
+		beforeEach(function() {
+			this.stub = sinon.stub(Build.prototype, 'doFileMerge');
+			this.stub2 = sinon.stub(Build.prototype, 'isValidFile');
```
Comment (rob-mccann) [inline test/buildscript-test.js]: :+1:

File: test/buildscript-test.js [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+		});
+		afterEach(function() {
+			Build.prototype.doFileMerge.restore();
```
Comment (rob-mccann) [inline test/buildscript-test.js]: the sandbox is your friend

File: test/buildscript-test.js (hunk 4 above)
Comment (rob-mccann) [inline test/buildscript-test.js]: would be good to call the variable `this.doFileMergeStub` for clarity

File: test/buildscript-test.js [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+			this.stub.returns(okConfigs.generic);
+			expect(build.isValidFile('')).to.eq(true);
+			this.stub.returns(okConfigs.finn);
+			expect(build.isValidFile('')).to.eq(true);
+			this.stub.returns(okConfigs.vg);
+			expect(build.isValidFile('')).to.eq(true);
+			this.stub.returns(okConfigs.bt);
+			expect(build.isValidFile('')).to.eq(true);
+		});
+		it('should throw when config is faulty', function() {
```
Comment (rob-mccann) [inline test/buildscript-test.js]: a bit too generic ;)

File: Gruntfile.js [hunk 7]
```diff
@@ -15,17 +15,29 @@ module.exports = function (grunt) {
...
+                    'integrationtest/**/*.js',
+                    'buildscript/**/*.js'
+                ]
             }
         },
         watch: {
             all: {
-                files: ['lib/**/*.js', 'test/**/*.js', 'autotracking/**/*.js', 'integrationtest/**/*.js'],
+                files: [
+                    'lib/**/*.js',
```
Comment (rob-mccann) [inline Gruntfile.js]: should technically be a separate PR but we can let it slide :)

File: buildscript/buildFiles.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+Build.prototype.isValidFile = function(fileName) {
+	var data = this.readAndParseFile(fileName);
+	var validate = new Validate(data, fileName);
+	validate.validateConfig();
+	return true;
+};
+
+Build.prototype.doFileMerge = function(fileName) {
```
Comment (rob-mccann) [inline buildscript/buildFiles.js]: do you mean concatenate rather than merge?

**Comment by rob-mccann** [comment]
Few other misc bits:

- PR title is a wee bit ambiguous
- mixed indentation in a few places - might be worth adding an editorconfig and doing a pass over it
- might be worth squashing the commits

File: test/buildscript-test.js (hunk 3 above)
Comment (magska) [inline test/buildscript-test.js]: Handy to have them available for all tests

File: buildscript/buildFiles.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+	Validate = require('./validate');
+
+function Build() {
+	this.configPath = './configs/';
+	this.fs = fs;
+	this.mkdirp = mkdirp;
+}
+
+Build.prototype.doBuild = function() {
+	var self = this;
```
Comment (rob-mccann) [inline buildscript/buildFiles.js]: use `bind`

File: buildscript/buildFiles.js (hunk 1 above)
Comment (rob-mccann) [inline buildscript/buildFiles.js]: :P but it saves us reinventing the wheel - plus they will already be tested so it's much less risk, liability and effort to reuse something already out there.

Take a look at `grunt-concat`, `grunt-wrap` and writing a custom grunt task for validation.

File: buildscript/buildFiles.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
+/* global process */
+'use strict';
+
+var fs = require('fs'),
+	mkdirp = require('mkdirp'),
```
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: var on every variable?

File: buildscript/buildFiles.js (hunk 9 above)
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: What's the rule here?

File: buildscript/buildFiles.js [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+			self.tryToCreateOutFolder();
+			self.getAllConfigFiles(this.fs.readdirSync(self.configPath));
+		} else {
+			console.log('path not found');
+		}
+	});
+};
+
+Build.prototype.getAllConfigFiles = function(fileNameArray) {
+	if (fileNameArray.length > 0) {
```
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: I'd invert both of these conditionals to remove all indentation.

File: buildscript/buildFiles.js (hunk 1 above)
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: document this as the main function?

File: buildscript/validate.js [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+		throw new Error('Config should be an array in ' + this.fileName);
+	}
+	if (config.length < 1) {
+		throw new Error('There should be at lease one object in config ' + this.fileName);
+	}
+};
+
+Validate.prototype.hasValidThrottles = function() {
+	var configs = this.config.config;
+	for (var i = 0; i < configs.length; i++) {
```
Comment (alexanderkjeldaas) [inline buildscript/validate.js]: map?

File: buildscript/validate.js [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+		var config = configs[i];
+		if (typeof config.throttle !== 'number'){
```
Comment (alexanderkjeldaas) [inline buildscript/validate.js]: isn't there a preconditions library ala java guava where one can say:

`assertNumber(config.throttle)` and `assertNumber(config.throttle, "Throttle needs to be a number")` ?

Comment (magska) [inline buildscript/validate.js]: Not built in to JS afaik, this is what various sites and stackoverflow threads says is the recommended way of checking.

File: test/buildscript-test.js (hunk 3 above)
Comment (rob-mccann) [inline test/buildscript-test.js]: are they all required? Also, what happens if blocket changes its config, does that matter?

Comment (magska) [inline test/buildscript-test.js]: These are just mock configs. So there are some that should be ok, and some that should fail. The real configs are stored in a separate folder as separate files.

File: buildscript/validate.js [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+/**
+ * Throws error if config.cis or config.datacollector is not a valid url as a string
+ */
+Validate.prototype.hasValidEndpoints = function() {
+	var configs = this.config.config;
+	var self = this;
+	configs.map(function(obj) {
+		if (!Type.is(obj.dataCollector, undefined)) {
+			var dc = obj.dataCollector;
+			if (!Type.is(dc, String)) {
```
Comment (alexanderkjeldaas) [inline buildscript/validate.js]: now this can be replaced with:

`assert(Type.is(dc, String), 'data-collector is not a string ' + self.fileName)`

**Comment by alexanderkjeldaas** [comment]
lgtm, not wanting to hold this back.

File: buildscript/validate.js (hunk 14 above)
Comment (rob-mccann) [inline buildscript/validate.js]: What advantage does `Type.is` give over `typeof dc === 'string'`?

Comment (rob-mccann) [inline buildscript/validate.js]: also, be wary of this http://stackoverflow.com/questions/203739/why-does-instanceof-return-false-for-some-literals

File: buildscript/validate.js [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+};
+
+/**
+ * Throws error if config.cis or config.datacollector is not a valid url as a string
+ */
+Validate.prototype.hasValidEndpoints = function() {
+	var configs = this.config.config;
+	var self = this;
+	configs.map(function(obj) {
+		if (!Type.is(obj.dataCollector, undefined)) {
```
Comment (rob-mccann) [inline buildscript/validate.js]: can we do a falsey check here?

File: test/buildscript-test.js (hunk 3 above)
Comment (rob-mccann) [inline test/buildscript-test.js]: Can we put these into a separate file to make the test file easier to read? maybe call it `fixtures.json`

File: Gruntfile.js [hunk 16]
```diff
@@ -15,17 +15,29 @@ module.exports = function (grunt) {
...
+                    'autotracking/**/*.js',
+                    'integrationtest/**/*.js',
+                    'buildscript/**/*.js'
+                ],
                 tasks: ['devbuild', 'lint', 'karma:unit:run']
             }
         },
         jscs: {
-            main: ['lib/**/*.js', 'autotracking/**/*.js'],
+            main: ['**/*.js', 'autotracking/**/*.js', 'buildscript/**/*.js'],
```
Comment (alexanderkjeldaas) [inline Gruntfile.js]: split into separate lines

File: buildscript/buildFiles.js (hunk 10 above)
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: indentation

File: buildscript/buildFiles.js [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+		}
+	}.bind(this));
+};
+
+/**
+ * Iterates through an array of config filenames and concatenates them with the tracking source
+ *
+ * @param {Array} fileNameArray
+ */
+Build.prototype.getAllConfigFiles = function(fileNameArray) {
```
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: would something like this be clearer?

```js
filenameArray.map(function (file) {
    console.log('building: ' + file);
    assert(this.isValidFile(file));
    this.doFileConcatenate(file);
}).bind(this));
}
```

File: buildscript/buildFiles.js [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+	if (fileNameArray.length <= 0) {
+		return;
+	}
+	var fileName = this.configPath + fileNameArray.pop();
+	console.log('building: ' + fileName);
+
+	if (!this.isValidFile(fileName)) {
```
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: seems like isValidFile can't fail.

File: buildscript/buildFiles.js [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+	this.fs.appendFileSync(buildFileName, this.readFile('./dist/autoTracker.min.js'));
+	this.fs.appendFileSync(buildFileName, 'var pulse2config = ' + this.readFile(fileName) + ';');
+};
+
+/**
+ * Attempts to create the output directoru
+ */
+Build.prototype.tryToCreateOutFolder = function() {
+	this.mkdirp('./out', function(err) {
+		if (err) {
```
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: `assert(!err, 'Error creating output dir: ' + err);` ?

File: buildscript/buildFiles.js [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+
+/**
+ * Changes first letter to upper case and all other letters to lowe rcase
+ *
+ * @param {String} text
+ * @returns lower case string with first letter in upper case
+ */
+Build.prototype.capitalizeFirstLetter = function(text) {
+	text = text.toLowerCase();
+    return text.charAt(0).toUpperCase() + text.slice(1);
```
Comment (alexanderkjeldaas) [inline buildscript/buildFiles.js]: indentation

File: buildscript/validate.js [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+};
+
+/**
+ * Throws error if the client property in config is not a lowercase a-z string.
+ */
+Validate.prototype.hasValidClient = function() {
+	var client = this.config.client;
+	assert(typeof client === 'string', 'Client ID not valid for ' + this.fileName);
+	if (!/^[a-z]+$/.test(client)) {
+		if (client !== '') {
```
Comment (alexanderkjeldaas) [inline buildscript/validate.js]: assert?

File: buildscript/validate.js [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+ * Checks if a URL is valid
+ *
+ * @param {String} url
+ * @returns bool
+ */
+Validate.prototype.isValidUrl = function(url) {
+	// jscs:disable
+	var pattern = /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g;
+	// jscs:enable
+	if (pattern.test(url)) {
```
Comment (alexanderkjeldaas) [inline buildscript/validate.js]: `return pattern.test(url);`

File: buildscript/validate.js (hunk 21 above)
Comment (magska) [inline buildscript/validate.js]: In this case, no, as there is more than one condition, I could write that in to the reqex, but that makes everything less readable.

**Comment by alexanderkjeldaas** [comment]
lgtm
